pkg/api/query/v1.go

```diff
@@ -1588,3 +1600,36 @@ func NewMetricMetadataHandler(client metadata.UnaryClient, enablePartialResponse
 		return t, warnings.AsErrors(), nil, func() {}
 	}
 }
+
+// MinResolutionFromExpr inspects the query expression and find the minimum required resolution for the
+// query. Right now it only considers promql functions that require at least 2 samples such as rate, irate, etc.
+func MinResolutionFromExpr(expr parser.Expr) time.Duration {
+	var minRes time.Duration
+	parser.Inspect(expr, func(n parser.Node, nodes []parser.Node) error {
+		switch e := n.(type) {
+		case *parser.MatrixSelector:
+			var found bool
+		OUTER:
+			for i := len(nodes) - 1; i >= 0; i-- {
+				node := nodes[i]
+				if f, ok := node.(*parser.Call); ok {
+					switch f.Func.Name {
+					// Functions that require at least 2 samples.
+					case "rate", "irate", "increase", "idelta", "deriv", "predict_linear":
```

Comment:
I'm 99% sure that we will forget to update this list once a new function appears that takes in a range vector. For instance, even right now this doesn't cover `xrate`, `xincrease`, `xdelta`. Perhaps this list could be converted into a list of functions like `map[string]bool` and then another test could be added that would check whether all `parser.Functions` are covered. This would make it so that it would be impossible to add another PromQL function without checking whether it needs to be added to this.